pwiz_tools/Shared/Common/Chemistry/Formula.cs

```diff
@@ -292,9 +353,78 @@ public static T ParseExpression(String formula)
                     }
                 }
             }
+
+            // Deal with mass modifications - combine if more than one, ensure positive count if only one
+            RegularizeMassModifications(result);
             return result;
         }
 
+        public static void AddMassModification(IDictionary<string, int> result, double modMass)
+        {
+            var massModStr = Molecule.FormatMassModification(modMass);
+            if (result.TryGetValue(massModStr, out var exist))
+            {
+                result[massModStr] = exist + 1;
+            }
+            else
+            {
+                result.Add(massModStr, 1);
+            }
+            RegularizeMassModifications(result); // Combine with any existing mass modifications
+        }
+
+        // Tidy up any mass modifications (e.g. [+2.45], [-3.55] etc) in the dictionary - combine if more than one, ensure positive count if only one
+        public static void RegularizeMassModifications(IDictionary<string, int> result)
+        {
+            var nMassMods = result.Keys.Count(k => k.StartsWith(MASS_MOD_START));
+            if (nMassMods == 1)
+            {
+                var kvp = result.First(k => k.Key.StartsWith(MASS_MOD_START));
+                if (kvp.Value < 0)
+                {
+                    // Just invert polarity
+                    var newKey = kvp.Key.Contains(@"+") ? kvp.Key.Replace(@"+", @"-") : kvp.Key.Replace(@"-", @"+");
+                    result.Add(newKey, -kvp.Value);
+                    result.Remove(kvp.Key);
+                }
+            }
+            else if (nMassMods > 1)
+            {
+                var massValueMono = 0.0;
+                var massValueAverage = 0.0;
+                var observedDecimals = 1; // Always show at least one decimal place
+                foreach (var kvp in result.Where(kvp => kvp.Key.StartsWith(MASS_MOD_START)).ToArray())
+                {
+                    var decimals = (kvp.Key.Length - kvp.Key.LastIndexOf('.')) - 2; // Omit . and ] in the count
+                    observedDecimals = Math.Max(decimals, observedDecimals);
+                    massValueMono += ParseMassModification(kvp.Key, false) * kvp.Value;
+                    massValueAverage += ParseMassModification(kvp.Key, true) * kvp.Value;
+                    result.Remove(kvp.Key);
+                }
+
+                var newMassMod = FormatMassModification(massValueMono, massValueAverage, observedDecimals);
+                result.Add(newMassMod, 1);
+            }
+        }
+
+        public bool HasMassModifications => Keys.Any(k => k.StartsWith(MASS_MOD_START));

```

Comment:
This does not belong in this class.
A "Molecule" should never have mass modifications on it. If you want a Molecule with mass modifications you should have to use the class "MoleculeMassOffset".
I think the code for parsing formulas that might have mass offsets in the middle of them should be confined to the stuff that deals with HardKlor. By the time the formula gets written to something like a .blib file, it should look like what MoleculeMassOffset.ToString() produces.

Reply:
Can't argue with that semantically - molecules are strictly speaking made of atoms and isotopes thereof, nothing more. 

But if we want to use a special class for the idea of "thing that results in an isotopic distribution", that would be a pretty extensive change to the codebase. I'd think that we'd wind up using that new class nearly everywhere that we currently use Molecule.

As to tolerating the offset description being randomly placed, consider this: if we understand that simply concatenating "C4H9O" and "H3O2N" is meaningful, then I'd think we'd want to understand that concatenating "C4H9O[+1.07]" and "H3O2N[-2.3]" is meaningful too.

Reply:
The class for "Thing that can result in an isotope distribution" is "MoleculeMassOffset".
"C4H9O[+1.07]" and "H3O2N[-2.3]" are both MoleculeMassOffset objects.
When you combine them you get another MoleculeMassOffset.

Molecules happen to have the property that if you concatenate their ToString() values, you get something that can still be parsed as a Molecule. We do not necessarily need to preserve that behavior for the class MoleculeMassOffset.

Reply:
I guess I'd have to take a shot at doing all the necessary replacements before I can pursue the cost/benefit argument. Brendan didn't think the current implementation was too nutty, at first blush anyway.

Reply:
I didn't look at the specifics, but I probably agree with Nick that it should be something other than a molecule. Maybe a "MoleculeRoughGuess". That's what it really is. We are just saying that probably the chemical formula we have is close enough to give us a good guess at the isotope distribution which will be dominated by the number of Carbon atoms in the molecule. So, whether those are exactly so many Ns, Os, and Hs and knowing that exactly is less important that knowing roughly the number of Carbon atoms and a highly precise center for the m/z distributions or centroids. Mass accuracy and precision are too high for us to just settle for a rough guess on the molecule mass, but for the isotope distribution a rough guess on the number of atoms, especially Carbons, may be just fine.

Reply:
Let me make one more argument for my current implementation then I'll drop it:

We already have a peculiar syntax for mass shifts in molecular formulas (the  ' and " thing), so it seems wrong to be hung up on one more like "[+1.1/1.12]". If we were going to be truly purist about Molecule/Formula we'd need to go back and revise everything to use standard isotope notation like 13C and 15N. But of course we'll never do that because the current notation is convenient for us and easy enough for users to understand, and it certainly wouldn't be worth the code churn. 

I don't think it's worth the code churn of replacing Formula and Molecule virtually everywhere they're used in Skyline for this reason either. (Looks like at least 88 and 140 places in the code, respectively)

Reply:
That is just a notation difference for specifying a thing that truly exists in nature. They have very clear definitions as molecules, even if the notation is less purist. They really are formulae and molecules.